File: SoloLearn Codes/Poker Hand/PokerHand.cs
using System;
using System.Collections.Generic;
using System.Data;
using System.Linq;

namespace SoloLearn
{
    class Program
    {
        static void Main(string[] args)
        {
            string[] hand = Console.ReadLine().Split();
            Dictionary<string, int> rankValues = new Dictionary<string, int>()
            {
                ["2"] = 2,
                ["3"] = 3,
                ["4"] = 4,
                ["5"] = 5,
                ["6"] = 6,
                ["7"] = 7,
                ["8"] = 8,
                ["9"] = 9,
                ["10"] = 10,
                ["J"] = 11,
                ["Q"] = 12,
                ["K"] = 13,
                ["A"] = 14
            };
            List<int> rank = hand
                .Select(x => rankValues[x.Substring(0, x.Length - 1)])
                .OrderByDescending(x => x).ToList();
            List<int> frequencies = rank.Select(x => rank.Count(y => x == y)).ToList();
            bool suitFrequency = hand.Select(x => x[x.Length - 1]).Distinct().Count() == 1;
            bool rankOrder = Enumerable.Range(1, 4).All(x => rank[x] + 1 == rank[x - 1]);

            if (suitFrequency && rankOrder && rank[0] == 14) Console.WriteLine("Royal Flush");
            else if (suitFrequency && rankOrder) Console.WriteLine("Straight Flush");
            else if (frequencies[0] == 4 || frequencies[4] == 4) Console.WriteLine("Four of a Kind");
            else if (new HashSet<int> { 2, 3 }
            .SetEquals(new List<int> { frequencies[4], frequencies[0] }))
                Console.WriteLine("Full House");
            else if (suitFrequency) Console.WriteLine("Flush");
            else if (rankOrder) Console.WriteLine("Straight");
            else if (frequencies[2] == 3) Console.WriteLine("Three of a Kind");
            else if (frequencies[1] == 2 && frequencies[3] == 2) Console.WriteLine("Two Pairs");
            else if (rank.Distinct().Where(x => rank.Count(y => y == x) == 2).Count() == 1)
                Console.WriteLine("One Pair");
            else Console.WriteLine("High Card");
        }
    }
}

File: SoloLearn Codes/Poker Hand/poker_hand.rb
rankVal = { "2"=> 2, "3"=> 3, "4"=> 4, "5"=> 5, "6"=> 6, "7"=> 7, "8"=> 8, "9"=> 9, "10"=> 10, "J"=> 11, "Q"=> 12, "K"=> 13, "A"=> 14 }
hand = gets.chomp.split
rank = hand.map{ |c| rankVal[c[0..-2]] }.sort.reverse
freq = rank.map{ |c| rank.count(c) }
suitFreq = hand.map{ |c| c[-1] }.uniq.length == 1
rankOrd = (1..4).all?{ |c| rank[c] + 1 == rank[c - 1] }

if suitFreq && rankOrd && rank[0] == 14
    puts("Royal Flush")
elsif suitFreq && rankOrd
    puts("Straight Flush")
elsif freq[0] == 4 || freq[4] == 4
    puts("Four of a Kind")
elsif [2, 3].sort == [freq[0], freq[4]].sort
    puts("Full House")
elsif suitFreq
    puts("Flush")
elsif rankOrd
    puts("Straight")
elsif freq[2] == 3
    puts("Three of a Kind")
elsif freq[1] == 2 && freq[3] == 2
    puts("Two Pairs")
elsif rank.uniq.select{ |c| rank.count(c) == 2 }.length == 1
    puts("One Pair")
else
    puts("High Card")
end